language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -565,6 +567,52 @@ writePackageIdTs dir pkgId =
     T.writeFileUtf8 (dir </> "packageId.ts") $ T.unlines
       ["export default '" <> unPackageId pkgId <> "';"]
 
+writeIndexTs :: FilePath -> [Module] -> IO ()
+writeIndexTs packageSrcDir modules =
+  T.writeFileUtf8 (packageSrcDir </> "index.ts") lines
+  where
+    lines :: T.Text
+    lines = T.unlines $ header <> concatMap entry modules
+
+    header :: [T.Text]
+    header =
+      [ "import {default as __packageId} from \"./packageId\""
+      , "export const packageId = __packageId"
+      ]
+
+    entry :: Module -> [T.Text]
+    entry mod
+      | null $ defDataTypes mod = []
+      | otherwise = importDecl var path <> exportDecl var parts
+      where
+        parts = unModuleName (moduleName mod)
+        var = "__" <> T.intercalate "_" parts
+        path =  "\"./" <> T.intercalate "/" parts <> "\""
+
+    importDecl :: T.Text -> T.Text -> [T.Text]
+    importDecl var path =
+      [ "/* eslint-disable @typescript-eslint/camelcase */"
+      , "import * as " <> var <> " from "  <> path
+      ]
+
+    exportDecl :: T.Text -> [T.Text] -> [T.Text]
+    exportDecl = go 0
+      where
+        spaces i = T.pack $ replicate i ' '
+
+        go :: Int -> T.Text -> [T.Text] -> [T.Text]
+        go indent var [m] = let ws = spaces indent in
+          [ ws <> "/* eslint-disable @typescript-eslint/no-unused-vars */"
+          , ws <> "export import " <> m <> " = " <> var
+          ]
+        go indent var (m : parts) = let ws = spaces indent in
+          [ ws <> "/* eslint-disable @typescript-eslint/no-namespace */"
```

Comment:
This might be showing my ignorance of typescript but why do we use namespaces instead of ES2015 module syntax? All the docs I can find on this claim that this is outdated and legacy stuff and ES2015 module syntax should be preferered.

Reply:
Each ES2015 module has to go into its own file, which is exactly the opposite of what we're trying to do here. However, _I think_ we can use proper `const`s and nested objects for the value level and `declare namspace` for the type level to achieve what we want. Let's try that in a separate PR though.

Reply:
Oh I see, I didn’t think of the type-level stuff so I thought a simple object would be sufficient.